src/java/picard/sam/markduplicates/MarkDuplicates.java

```diff
@@ -443,11 +516,20 @@ private void generateDuplicateIndexes() {
         this.duplicateIndexes.doneAddingStartIteration();
     }
 
-    private boolean areComparableForDuplicates(final ReadEndsForMarkDuplicates lhs, final ReadEndsForMarkDuplicates rhs, final boolean compareRead2) {
-        boolean retval = lhs.libraryId == rhs.libraryId &&
-                lhs.read1ReferenceIndex == rhs.read1ReferenceIndex &&
-                lhs.read1Coordinate == rhs.read1Coordinate &&
-                lhs.orientation == rhs.orientation;
+    private boolean areComparableForDuplicates(final ReadEndsForMarkDuplicates lhs, final ReadEndsForMarkDuplicates rhs, final boolean compareRead2, final boolean useBarcodes) {
+        boolean retval = lhs.libraryId == rhs.libraryId;
+
+        if (retval && useBarcodes) {
+            if (retval) retval = 0 == compareInteger(((ReadEndsForMarkDuplicatesWithBarcodes) lhs).barcode, ((ReadEndsForMarkDuplicatesWithBarcodes) rhs).barcode);
+            if (retval) retval = 0 == compareInteger(((ReadEndsForMarkDuplicatesWithBarcodes) lhs).readOneBarcode, ((ReadEndsForMarkDuplicatesWithBarcodes) rhs).readOneBarcode);
+            if (retval) retval = 0 == compareInteger(((ReadEndsForMarkDuplicatesWithBarcodes) lhs).readTwoBarcode, ((ReadEndsForMarkDuplicatesWithBarcodes) rhs).readTwoBarcode);
+        }
+
+        if (retval) {
+            retval = lhs.read1ReferenceIndex == rhs.read1ReferenceIndex &&
+                    lhs.read1Coordinate == rhs.read1Coordinate &&
+                    lhs.orientation == rhs.orientation;
+        }
```

Comment:
It might be worth renaming ```retval``` to something else (e.g. ```areComparable```). It's confusing (and possibly bug-prone) to switch between this code where it means one thing, and ```ReadEndsMDComparator.compare``` below where it means something different.